File: src/cli/types/jobmanagertypes/task.go
package jobmanagertypes

import (
    "strconv"
    "sync"
    "sync/atomic"
    "time"

    "google.golang.org/protobuf/types/known/timestamppb"

    "github.com/awslabs/filemoverexpress/logger"
    "github.com/awslabs/filemoverexpress/types/pbtypes/fme/v1"
    "github.com/awslabs/filemoverexpress/utils"
    "github.com/awslabs/filemoverexpress/utils/safeconv"
)

const (
    TaskStatusPreFlight    TaskStatus = "PRE-FLIGHT"
    TaskStatusQueued       TaskStatus = "QUEUED"
    TaskStatusInProgress   TaskStatus = "IN_PROGRESS"
    TaskStatusCompleted    TaskStatus = "COMPLETED"
    TaskStatusChecksumming TaskStatus = "CHECKSUMMING"
    TaskStatusCancelled    TaskStatus = "CANCELLED"
    TaskStatusPaused       TaskStatus = "PAUSED"
    TaskStatusSkipped      TaskStatus = "SKIPPED"
    TaskStatusError        TaskStatus = "ERROR"

    TaskDirectionUpload   TaskDirection = "UPLOAD"
    TaskDirectionDownload TaskDirection = "DOWNLOAD"
)

type (
    TaskStatus    string
    TaskDirection string

    LocalFile struct {
        LastModified time.Time
        Path         string
        Size         int64
    }

    S3Object struct {
        Key          string
        LastModified time.Time
        Size         int64
    }

    Task struct {
        destination      string
        taskId           string
        localFile        LocalFile
        s3Object         S3Object
        taskDirection    TaskDirection
        status           TaskStatus
        statusMessage    string
        jobId            string
        checksum         string
        lock             *sync.RWMutex
        priority         int
        err              error
        BytesTransferred int64
    }

    TaskConfig struct {
        Destination   string
        LocalFile     LocalFile
        S3Object      S3Object
        JobId         string
        TaskDirection TaskDirection
        Priority      int
    }
)

func NewTask(config TaskConfig) (*Task, error) {
    taskId, err := utils.CalculateTransferId(config.S3Object.Key, config.LocalFile.Path, strconv.Itoa(time.Now().Nanosecond()))
    if err != nil {
        return nil, err
    }
    return &Task{
        taskId:        taskId,
        localFile:     config.LocalFile,
        s3Object:      config.S3Object,
        status:        TaskStatusPreFlight,
        jobId:         config.JobId,
        taskDirection: config.TaskDirection,
        priority:      config.Priority,
        destination:   config.Destination,
        lock:          &sync.RWMutex{},
    }, nil
}

// region Getters and Setters

func (t *Task) GetSourcePath() string {
    if t.TaskDirection() == TaskDirectionDownload {
        return t.S3Object().Key
    }
    return t.LocalFile().Path
}

func (t *Task) GetSize() int64 {
    if t.TaskDirection() == TaskDirectionDownload {
        return t.S3Object().Size
    }
    return t.LocalFile().Size
}

func (t *Task) TaskId() string {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.taskId
}

func (t *Task) SetTaskId(taskId string) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.taskId = taskId
}

func (t *Task) Destination() string {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.destination
}

func (t *Task) SetDestination(destination string) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.taskId = destination
}

func (t *Task) TaskDirection() TaskDirection {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.taskDirection
}

func (t *Task) SetTaskDirection(taskDirection TaskDirection) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.taskDirection = taskDirection
}

func (t *Task) JobId() string {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.jobId
}

func (t *Task) SetJobId(jobId string) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.jobId = jobId
}

func (t *Task) S3Object() S3Object {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.s3Object
}

func (t *Task) SetS3Object(s3Object S3Object) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.s3Object = s3Object
}

func (t *Task) Status() TaskStatus {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.status
}

func (t *Task) SetStatus(status TaskStatus) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.status = status
}

func (t *Task) SetStatusAndError(status TaskStatus, err error) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.status = status
    t.err = err
}

func (t *Task) Err() error {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.err
}

func (t *Task) SetErr(err error) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.err = err
}

func (t *Task) StatusMessage() string {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.statusMessage
}

func (t *Task) SetStatusMessage(statusMessage string) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.statusMessage = statusMessage
}

func (t *Task) LocalFile() LocalFile {
    t.lock.RLock()
    defer t.lock.RUnlock()
    return t.localFile
}

func (t *Task) SetLocalFile(localFile LocalFile) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.localFile = localFile
}

func (t *Task) Priority() int {
    t.lock.RLock()
    defer t.lock.RUnlock()
    if t.status == TaskStatusPaused {
        return t.priority * -1
    }
    return t.priority
}

func (t *Task) SetPriority(priority int) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.priority = priority
}

func (t *Task) Checksum() string {
    t.lock.Lock()
    defer t.lock.Unlock()
    return t.checksum
}

func (t *Task) SetChecksum(checksum string) {
    t.lock.Lock()
    defer t.lock.Unlock()
    t.checksum = checksum
}

func (t *Task) ToProtobuf() *fmev1.Task {
    t.lock.RLock()
    defer t.lock.RUnlock()

    var errorString string
    if t.err != nil {
        errorString = t.err.Error()
    }

    // Safe conversion for priority - Issue #8
    priority, err := safeconv.IntToInt32(t.priority)
    if err != nil {
        logger.Error("Invalid Priority value %d for task %s: %v, using default", t.priority, t.taskId, err)
        priority = 0 // Default priority
    }

    lf := fmev1.TaskLocalFile{
        Path:         t.localFile.Path,
        Size:         t.localFile.Size,
        LastModified: timestamppb.New(t.localFile.LastModified),
    }

    s3o := fmev1.TaskS3Object{
        Key:          t.s3Object.Key,
        Size:         t.s3Object.Size,
        LastModified: timestamppb.New(t.s3Object.LastModified),
    }

    return &fmev1.Task{
        TaskId:           t.taskId,
        Destination:      t.destination,
        LocalFile:        &lf,
        S3Object:         &s3o,
        TaskDirection:    string(t.taskDirection),
        Status:           string(t.status),
        StatusMessage:    t.statusMessage,
        JobId:            t.jobId,
        Checksum:         t.checksum,
        Priority:         priority,
        Err:              errorString,
        BytesTransferred: atomic.LoadInt64(&t.BytesTransferred),
    }
}

//endregion

File: src/gui/src/app/classes/grpc/task.ts
import { cleanPath } from '@app/components/layout/file-browser/file-browser.utils';
import { Task as ProtoTask, TaskLocalFile as ProtoLocalFile, TaskS3Object as ProtoS3Object } from '@gen/es/fme/v1/job_pb';
import { timestampDate } from '@bufbuild/protobuf/wkt';

export interface FlattenedTask {
    taskId: string,
    destination: string,
    source: string,
    size: number,
    lastModified: Date | null,
    direction: string,
    status: string,
    statusMessage: string,
    jobId: string,
    checksum: string,
    priority: number,
    error: string,
    bytesTransferred: number,
}

export class Task {
    constructor(
        public taskId: string,
        public destination: string,
        public localFile: LocalFile,
        public s3Object: S3Object,
        public direction: string,
        public status: string,
        public statusMessage: string,
        public jobId: string,
        public checksum: string,
        public priority: number,
        public error: string,
        public bytesTransferred: number,
    ) {
    }

    static fromProtobuf(task: ProtoTask): Task {
        if (!task.localFile || !task.s3Object) {
            throw new Error('Missing required data in message');
        }

        return new Task(
            task.taskId,
            task.destination,
            LocalFile.fromProtobuf(task.localFile),
            S3Object.fromProtobuf(task.s3Object),
            task.taskDirection,
            task.status,
            task.statusMessage,
            task.jobId,
            task.checksum,
            task.priority,
            task.err,
            Number(task.bytesTransferred),
        );
    }

    static toFlattenedTask(task: Task, bucket: string): FlattenedTask {
        let destination = task.destination;
        let source = '';
        let size = 0;
        let lastModified = null;

        switch (task.direction.toLowerCase()) {
            case 'upload':
                destination = bucket ? `s3://${bucket}/${cleanPath(task.destination)}` : task.destination;
                source = task.localFile.path;
                size = task.localFile.size;
                lastModified = task.localFile.lastModified;
                break;
            case 'download':
                destination = task.destination;
                source = bucket ? `s3://${bucket}/${cleanPath(task.s3Object.key)}` : task.s3Object.key;
                size = task.s3Object.size;
                lastModified = task.s3Object.lastModified;
        }

        return {
            taskId: task.taskId,
            destination: destination,
            source: source,
            size: size,
            lastModified: lastModified,
            direction: task.direction,
            status: task.status,
            statusMessage: task.statusMessage,
            jobId: task.jobId,
            checksum: task.checksum,
            priority: task.priority,
            error: task.error,
            bytesTransferred: task.bytesTransferred,
        };
    }
}

class LocalFile {
    constructor(public path: string, public size: number, public lastModified: Date) {
    }

    static fromProtobuf(lf: ProtoLocalFile) {
        return new LocalFile(
            lf.path,
            Number(lf.size),
            lf.lastModified ? timestampDate(lf.lastModified) : new Date(),
        );
    }
}

class S3Object {
    constructor(public key: string, public size: number, public lastModified: Date) {
    }

    static fromProtobuf(s3o: ProtoS3Object): S3Object {
        return new S3Object(
            s3o.key,
            Number(s3o.size),
            s3o.lastModified ? timestampDate(s3o.lastModified) : new Date(),
        );
    }
}
